header/p2p/peer_tracker.go

```diff
@@ -102,3 +140,50 @@ func (p *peerTracker) peers() []*peerStat {
 	}
 	return peers
 }
+
+// gc goes through connected and disconnected peers once in gcPeriod
+// and removes every peer that meets conditions:
+// * disconnected peer will be removed if it is being disconnected for more than maxAwaitingTime;
+// * connected peer will be removed if it scores less or equal than defaultScore;
+func (p *peerTracker) gc() {
+	ticker := time.NewTicker(p.gcPeriod)
+	for {
+		select {
+		case <-p.ctx.Done():
+			p.done <- struct{}{}
+			return
+		case <-ticker.C:
+			wg := sync.WaitGroup{}
+			p.Lock()
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				for id, peer := range p.disconnectedPeers {
+					if peer.removedAt.Before(time.Now()) {
+						delete(p.disconnectedPeers, id)
+					}
+				}
+			}()
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				for id, peer := range p.connectedPeers {
+					if peer.peerScore <= p.defaultScore {
+						delete(p.connectedPeers, id)
+					}
+				}
+			}()
+			wg.Wait()
+			p.Unlock()
```

Comment:
? why does this have to be async?

Reply:
I made this change before I realised the we should have peerTracker limit size. And probably we could have 100000 of peers in cache, so it was added to speed gc(as we are locking the mutex). But we do not need them to be async.

Reply:
https://github.com/celestiaorg/celestia-node/pull/1298/commits/74283254d13527d615d6c57b70e6e62846413ad1